### PR TITLE
Issue 5149 User Permission - Warning modal when changing permissions

### LINKED ISSUE #5149: User Permissions - Warning modal when making a change
### Description
Product issue: https://github.com/3drepo/3D-Repo-Product-Team/issues/537
Be warned when any user permissions are changed.

### Goals
- [ ] As a teamspace/project admin, show me a modal that requires confirmation before changing a single or multiple users permissions with information regarding the change.
- [ ] As a teamspace/project admin, allow me to suppress this modal on each session by adjusting the toggle option in the user permissions UI.

### Tasks
- [ ] Add a boolean switch in users permissions page (design is just a guide for this)
- [ ] Use local storage to save the switch permission settings ( like with visual settings)
- [ ] Dispatch a warning message like in the design whenever the user is changing something)  

### Related Resources
[MockUp](https://www.figma.com/design/eI1KkbqCaxI7AeMeXnPE1V/Permissions?node-id=1-38)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #5149

#### Description
Changing a user permissions prompts a modal if the flag is set to false. 
It is possible to toggle the flag at any point from the pages where you can change permissions

#### Test cases
- Try to change the permissions from:
   - Users tab in teamspace page;
   - Project permission tabs in project page (both 1 or a batch);
   - Users permissions tabs in project page; (both 1 or a batch);

- Make sure setting the modal flag to false will not trigger the modal
- Make sure, after logging out and logging back in, the flag is set to false

### CODE REVIEW

File: frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx [hunk 1]
```diff
@@ -91,8 +93,13 @@ export class CellSelect extends PureComponent<IProps, IState> {
 		const selectedValue = event.target.value;
 
 		if (this.state.selectedValue !== selectedValue) {
-			this.setState({selectedValue});
-			this.props.onChange(event, selectedValue);
+			updatePermissionsOrTriggerModal({
```
Comment (sanmont3drepo) [inline frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx:96]: I dont think this is the best place to add the permissions confimation. cellSelect seems to be used widely in other places. foe example when changin the risks properties:
cellselectproblemst

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/test/mocks/react-children-utilities.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/**
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/test/mocks/react-children-utilities.js:1]: Unfortunately, this is due to Jest (indirectly) importing a library called `react-children-utilities`. The latter probably uses CommonJS, which relies on a different export system, and Jest does not recognise it and triggers the error `SyntaxError: Cannot use import statement outside a module`

File: frontend/src/v4/routes/components/permissionsTable/permissionsTable.component.tsx [hunk 3]
```diff
@@ -110,7 +112,7 @@ interface IProps {
 	context?: string;
 	onSelectionChange?: (selectedUsers) => void;
 	onFilterChange?: (filteredUsers) => void;
-	onPermissionsChange?: (permissions) => void;
+	onPermissionsChange?: (permissions, permissionsType, permissionsCount?) => void;
```
Comment (sanmont3drepo) [inline frontend/src/v4/routes/components/permissionsTable/permissionsTable.component.tsx:115]: dont change base components that are being used in other parts of the system (like in invitationdialogs)

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx [hunk 4]
```diff
@@ -91,7 +91,6 @@ export class CellSelect extends PureComponent<IProps, IState> {
 		const selectedValue = event.target.value;
 
 		if (this.state.selectedValue !== selectedValue) {
```
Comment (sanmont3drepo) [inline frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx:93]: im worried about the deleted line here. why was this deleted?

Comment (Amantini1997) [reply frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx:93]: Imagine this scenario:
image
The user has the value set to `Teamspace admin` and attempts to change it to `User`.
The confirmation modal is prompted, and if they cancel the action from the modal, the value should still be `Teamspace admin`, but if this line is not deleted, the shown value is `User` instead.

So I rely on the `componentDidUpdate` to update the state instead

Comment (sanmont3drepo) [reply frontend/src/v4/routes/components/customTable/components/cellSelect/cellSelect.component.tsx:93]: I think we can get rid of the whole `selectedValue` state because that seems intended for uncontrolled inputs and we are not using those types of inputs in in v4

**Review by sanmont3drepo — APPROVED** [review]

**Review by sanmont3drepo — DISMISSED** [review]

**Review by carmenfan — CHANGES_REQUESTED** [review]
as discussed, let's remove the confirmation modal logic from the users page
image

**Review by sanmont3drepo — APPROVED** [review]

File: frontend/src/v4/modules/userManagement/userManagement.sagas.ts [hunk 5]
```diff
@@ -282,6 +286,13 @@ export function* fetchProject({ project }) {
 
 export function* updateProjectPermissions({ permissions }) {
 	try {
+		const shouldUpdate = yield updatePermissionsOrTriggerModal({
```
Comment (sanmont3drepo) [inline frontend/src/v4/modules/userManagement/userManagement.sagas.ts:289]: it seems that updating the permissions fails
failed

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by carmenfan — APPROVED** [review]
